internal/util/sizeof.go

```diff
@@ -0,0 +1,297 @@
+// Copyright 2024 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package util
+
+import (
+	"reflect"
+
+	"github.com/googlecloudplatform/gcsfuse/internal/storage/gcs"
+	"google.golang.org/api/googleapi"
+	storagev1 "google.golang.org/api/storage/v1"
+)
+
+var (
+	// pointerSize represents the size of the pointer of any type.
+	pointerSize int
+)
+
+func init() {
+	var i int
+	pointerSize = int(reflect.TypeOf(&i).Size())
+}
+
+// Definitions/conventions (not based on a standard, but just made up for convenience).
+//   1. Raw/unsafe size: This is the size of empty
+//      data structures, i.e. size of memory, for data
+//      structures when just initialized, without any
+//      content filled into them.
+//      This is same as what is returned by unsafe.Sizeof(...)
+//
+//      Built-in types ([u]int*, bool, time.*) have the same
+//      same raw/unsafe size as their actual space in memory.
+//
+//      More complex types like string, array/slice have a fixed
+//      /standard raw/unsafe size e.g. string is 16 bytes (8 for
+//      holding pointer to its content, 8 bytes for holding length).
+//      But it doesn't account for string's content, so any string,
+//  	be it emoty-string, or "hello" or a million-char string, all
+//      have Raw/unsafe size of 16.
+//
+//      Pointers have a raw/unsafe size of 8 bytes on 64-bit platforms.
+//
+//      Similarly, an array/slice/map etc. have small fixed Raw/unsafe size like 16 or 32 bytes.
+//
+//      Raw/unsafe size of a struct is the sum of Raw/unsafe sizes of all its constituents.
+//
+//   2. Content-size (or nested-content-size): This is the additional size added when some content is
```

Comment:
I think, nestedSize and nestedContentSize both together hard to follow, specially when a new terminology content size comes into the picture.

We can chose only one and defined their behavior for struct, slices and maps should be enough.

Reply:
Not sure I follow the point. PLease clarify.

Reply:
@raj-prince please respond.

Reply:
I was saying, content and nested-content creates confusion. Hence, only content-size will be okay. As you have already defined what does content size means for structs, slices and maps.

Reply:
Removed the name nested-content-size.